Add native XPU support for `linalg.qr`, `geqrf`, and `householder_product`

mwiktor-intel opened the pull request:
Implement `torch.linalg.qr`, `torch.geqrf`, and `torch.linalg.householder_product` on XPU by registering `geqrf_stub` and `orgqr_stub`, which allows reusing PyTorch's existing QR decomposition logic.

Implementation details:
- `orgqr`/`ungqr` (Q matrix recovery): uses oneMKL LAPACK when built with `USE_ONEMKL_XPU`, with CPU fallback otherwise. Uses the batched API (`orgqr_batch`/`ungqr_batch`) when `k == n`, and a per-item loop otherwise
- `geqrf` (QR factorization): uses CPU fallback for now (will be switch to oneMKL once XPU performance is optimized)
- supports both real floating-point and complex types

Changes:
- Register `geqrf_stub` and `orgqr_stub` XPU dispatchers in `BatchLinearAlgebra.cpp`
- Add `orgqr_mkl` and `ungqr_mkl` oneMKL kernel implementations in `xpu/mkl/BatchLinearAlgebra.cpp`
- Register `linalg_qr.out`, `linalg_householder_product{,.out}`, and `geqrf{,.a}` ops for XPU in `native_functions.yaml`
- Remove `geqrf`, `linalg_householder_product`, and `linalg_qr.out` from the XPU CPU-fallback list
- Add XPU-specific tests for QR modes (`r, reduced, complete`) covering batched inputs, various dimensions, and float/complex dtypes

Fixes https://github.com/intel/torch-xpu-ops/issues/1900

Silv3S committed:
Implement linalg.qr on XPU via `geqrf` and `orgqr` stubs

Add `geqrf` and `orgqr` kernel implementations for XPU using oneMKL:
- Register `geqrf_stub` with CPU fallback path (oneMKL `geqrf`/`geqrf_batch`
  underperforms vs MKL on CPU; using CPU fallback yields 2-4x
  end-to-end speedup for `linalg.qr`)
- Register `orgqr_stub` with native XPU execution via oneMKL
- Support both real (`orgqr`) and complex (`ungqr`) dtypes
- Support batched inputs via oneMKL `geqrf_batch`/`ungqr_batch`
- Use `at::xpu::getCurrentSYCLQueue()` for device execution
- Use `at::empty` instead of `malloc_device` and `free`
- Early return on empty tensors

The `linalg.qr` op dispatches through these stubs, combining CPU `geqrf`
with XPU `orgqr` for optimal end-to-end performance.

CuiYifeng reviewed (changes requested):
Please check comments and fix related failed cases.

CuiYifeng commented on yaml/native/native_functions.yaml:
```diff
@@ -9443,6 +9443,17 @@
+  structured: True
+  dispatch:
+    XPU: linalg_qr_xpu_out
+
```
@mwiktor-intel For layout differences, could you indicate in MKL document the reason why LAPACK is row-major? https://www.intel.com/content/www/us/en/docs/onemkl/developer-reference-dpcpp/2025-2/geqrf-usm-version.html Please pay attention to leading dimension `lda` of input matrix. Suppose matrix A is col-major storage and shape is (m, n), we can set `lda=m` to indicate correct layout.

mwiktor-intel replied:
LDA cannot be replacement of stride. LDA can be used to skip some elements in matrix, but with LDA we cannot force difference data orientation. Specifically: say we have 3x2 matrix. Lapack expects this 6 element array in memory to represent the data as follows:
[ 0 2 4
  1 3 5]
LDA here is 2, Setting LDA 3 does not change anything here The purpose of LDA is when, say, we have for any reason some bigger structure in memory, say,
 0  3 6 
1  4 7
2 5 8 ]
and we want to take only 2 upper rows to process. LDA here should be 3, giving information to the function, that apart from logical number of rows equal to 2, the procedure should skip 3 values to get to the data for 2nd column.

Some BLAS function have row_major of col_major order, but it is not the case within LAPACK.
Note, in the documentation, LDA cannot be set to less than number of rows in the processed matrix.

CuiYifeng commented on src/ATen/native/xpu/mkl/BatchLinearAlgebra.cpp:
```diff
@@ -561,4 +561,133 @@ void lu_factor_mkl(
+    b=1;
+    for (int dimension=0; dimension<range-2; dimension++) {
+      b*=a_contig.sizes().at(dimension);
+    }
```
`at::native::batchCount` can help you get flattened batch size. https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/native/LinearAlgebraUtils.h#L112

CuiYifeng reviewed (changes requested):
Please check comments (including Copilot's). This PR needs to be reviewed again after fixing a series of failed cases.
Suggest to refresh code format before review. Furthermore, please also check performance compared to CPU.

pbielak commented on src/ATen/native/xpu/mkl/BatchLinearAlgebra.cpp:
```diff
@@ -561,4 +561,133 @@ void lu_factor_mkl(
+  int64_t b = numel ==0 ? 0 : numel / mn;
+
+
+  if (b==0 && mode=="complete" && n>0) {
```
Fixed

pbielak commented on src/ATen/native/xpu/mkl/BatchLinearAlgebra.cpp:
```diff
@@ -571,4 +571,143 @@ void lu_factor_mkl(
+
+  // normal case, non-zero dimensions
+  if (mode!="r") {
+    result_q.transpose_(-2, -1).contiguous();
```
Fixed

pbielak commented:
Initially I refactored the base implementation from @mwiktor-intel, but finally I ended up registering the `geqrf` and `orgqr` stubs, which allows to reuse the QR implementation from PyTorch. There is no observable performance gap between these two implementations, so I decided to remove Michał's implementation and use the function stubs. 

As oneMKL supports also complex dtypes for these kernels, I enabled the call path here too, while updating the tests to include complex dtypes as well.

Please review again @CuiYifeng

CuiYifeng commented on src/ATen/native/xpu/XPUFallback.template:
```diff
@@ -230,23 +230,18 @@ TORCH_LIBRARY_IMPL(aten, XPU, m) {
     "linalg_eigvals.out",
     "_linalg_eigh.eigenvalues",
-    "linalg_householder_product",
-    "linalg_householder_product.out",
```
Please also add `geqrf`, `geqrf.a`, `linalg_householder_product` and `linalg_householder_product.out` into `yaml/native/native_functions.yaml`.

pbielak replied:
Added the entries to the `native_functions.yaml`

CuiYifeng commented on src/ATen/native/xpu/mkl/BatchLinearAlgebra.cpp:
```diff
@@ -479,4 +479,156 @@ void triangular_solve_mkl(
+    return result;
+  }
+
+  auto& queue = c10::xpu::getCurrentXPUStream().queue();
```
```suggestion
  auto& queue = at::xpu::getCurrentSYCLQueue();
```

pbielak replied:
Applied

CuiYifeng commented on src/ATen/native/xpu/mkl/BatchLinearAlgebra.cpp:
```diff
@@ -479,4 +479,156 @@ void triangular_solve_mkl(
+    return result;
+  }
+
+  auto& queue = c10::xpu::getCurrentXPUStream().queue();
```
```suggestion
  auto& queue = at::xpu::getCurrentSYCLQueue();
```

pbielak replied:
Applied

CuiYifeng commented on src/ATen/native/xpu/mkl/BatchLinearAlgebra.cpp:
```diff
@@ -479,4 +479,156 @@ void triangular_solve_mkl(
+              lda,
+              std::max<int64_t>(1, matrix_stride),
+              batch_tau_stride,
+              std::max<int64_t>(1, batch_size));
```
With the early-return of `result.numel() == 0`, why can we meet batch_size < 1 here?

pbielak replied:
Must have overlooked after adding the early-return - removed the `std::max` calls

CuiYifeng commented on src/ATen/native/xpu/mkl/BatchLinearAlgebra.cpp:
```diff
@@ -479,4 +479,156 @@ void triangular_solve_mkl(
+              lda,
+              std::max<int64_t>(1, matrix_stride),
+              batch_tau_stride,
+              std::max<int64_t>(1, batch_size));
```
Ditto.

pbielak replied:
Removed the `std::max` calls

CuiYifeng commented on src/ATen/native/xpu/BatchLinearAlgebra.cpp:
```diff
@@ -134,4 +137,163 @@ void triangular_solve_kernel_xpu(
+  m.impl(
+      "linalg_householder_product.out",
+      TORCH_FN(at::native::linalg_householder_product_out_xpu));
+}
```
Since `geqrf_stub` and `orgqr_stub` have been registered for XPU kernel, why not reuse `linalg_householder_product_out`, `geqrf` and `geqrf_out` instead of using `TORCH_LIBRARY_IMPL`?

pbielak replied:
Done

CuiYifeng reviewed (changes requested):
@pbielak Please also resolve the code conflict and check the following new failed cases:

- op_ut,third_party.torch-xpu-ops.test.xpu.test_ops_xpu.TestCommonXPU,test_out_warning_geqrf_xpu
- op_ut,third_party.torch-xpu-ops.test.xpu.test_ops_xpu.TestCompositeComplianceXPU,test_cow_input_linalg_householder_product_xpu_float32

BBBela commented on src/ATen/native/xpu/mkl/BatchLinearAlgebra.cpp:
```diff
@@ -479,4 +479,142 @@ void triangular_solve_mkl(
+  const int64_t m = result.size(-2);
+  const int64_t n = result.size(-1);
+  const int64_t k = tau.size(-1);
+  const int64_t lda = std::max<int64_t>(1, m);
```
It looks, like the `std::max<int64_t>(1, m);` is always equal to `m` in this line. If `m` is zero, then `result.numel() == 0` is true and function returns early. Just for consideration, can stay as it is.

pbielak replied:
Right - I simplified this line

BBBela commented on src/ATen/native/xpu/mkl/BatchLinearAlgebra.cpp:
```diff
@@ -479,4 +479,142 @@ void triangular_solve_mkl(
+  const int64_t lda = std::max<int64_t>(1, m);
+  const int64_t batch_size = native::batchCount(result);
+  const int64_t matrix_stride = native::matrixStride(result);
+  const int64_t tau_stride = std::max<int64_t>(1, k);
```
It looks, like the `std::max<int64_t>(1, k);` is not necessary. It should work well when `k == tau_stride == 0` as mkl doesn't use the calculated `tau_batch_ptr` if `k == 0` anyway. Just for consideration, can stay as it is.

pbielak replied:
Right - I simplified this line

BBBela commented on src/ATen/native/xpu/mkl/BatchLinearAlgebra.cpp:
```diff
@@ -479,4 +479,142 @@ void triangular_solve_mkl(
+    const auto* tau_data = reinterpret_cast<const T*>(tau.const_data_ptr());
+
+    if (k == n) {
+      const int64_t batch_tau_stride = std::max<int64_t>(1, std::min(m, n));
```
Here also the `std::max<int64_t>` seems not necessarily required as the `m` or `n` equal `0` case would return early. Just for consideration, can stay as it is.

pbielak replied:
Right - I simplified this line

BBBela reviewed (commented):
Generally it looks good to me.

I have just three minor comments below, but fell free to resolve them if you don't find them relevant as they are not of high importance. Just wanted to point them for consideration.

I have just one bigger question - do we need all the newly added test cases? I mean, the operators implemented by this PR are already used in PyTorch by at least CPU and CUDA, so I suppose that they are tested in the upstream for correctness, aren't they? Does those test cases added in this PR perform some additional or more extensive checks that are not performed in the upstream? Aren't they duplications?

pbielak commented:
> Generally it looks good to me.
> 
> I have just three minor comments below, but fell free to resolve them if you don't find them relevant as they are not of high importance. Just wanted to point them for consideration.
> 
> I have just one bigger question - do we need all the newly added test cases? I mean, the operators implemented by this PR are already used in PyTorch by at least CPU and CUDA, so I suppose that they are tested in the upstream for correctness, aren't they? Does those test cases added in this PR perform some additional or more extensive checks that are not performed in the upstream? Aren't they duplications?

@BBBela I did not inspect these tests as they were added in the initial PR draft by @mwiktor-intel However, now I reevaluated their scope and they actually analogous cases as the upstream tests. Hence, I removed the tests added by this PR, as all the cases are covered by upstream tests. 

Thanks for all the review comments!

BBBela reviewed (approved):
Looks good to me, thanks!

pbielak committed:
Revert fallback removals and yaml changes (will be in separate PRs)

pbielak commented:
Due to updated build process (`native_functions.yaml` is now only available in PyTorch), this commit only provides the XPU implementations for geqrf oraz orgqr/ungqr. After merging, we need to land https://github.com/pytorch/pytorch/pull/187564 and finally the PR for removing the entries from the XPUFallback.cpp file https://github.com/intel/torch-xpu-ops/pull/4043 (so we can test the changes)

pbielak committed:
Use zero-initialized scratchpad for orgqr/ungqr to avoid MKL corruption

oneMKL's GPU orgqr kernel implicitly requires a zero-initialized
scratchpad. When the caching allocator reuses memory containing stale
data from prior operations, the computation may produce incorrect
results. Replace `at::empty` with `at::zeros` for all orgqr/ungqr
scratchpad allocations. No measurable performance regression.

pbielak commented:
@CuiYifeng @BBBela I had to change `at::empty` to `at::zeros` for scratchpad allocation, as in another test we found a regression, where re-using memory (provided via the XPU caching allocator), caused the QR kernel to return wrong values. After the inspection of the MKL code, I found that the memory must be zero-initialized.

Re-running the benchmark showed no performance regressions when using zeros instead of empty.

CuiYifeng commented:
> @CuiYifeng @BBBela I had to change `at::empty` to `at::zeros` for scratchpad allocation, as in another test we found a regression, where re-using memory (provided via the XPU caching allocator), caused the QR kernel to return wrong values. After the inspection of the MKL code, I found that the memory must be zero-initialized.
> 
> Re-running the benchmark showed no performance regressions when using zeros instead of empty.

Thanks for the elaboration. May I know in which test we found the regression in?

pbielak commented:
@CuiYifeng The test case: `test_linalg_xpu.py::TestLinalgXPU::test_det_logdet_slogdet_xpu_float64`. I reported the observation to MKL and it turned out to be a but, but it should have been fixed by now. Nonetheless, I would keep `at::zeros` as it doesn't introduce any noticeable overhead.

CuiYifeng commented:
> @CuiYifeng The test case: `test_linalg_xpu.py::TestLinalgXPU::test_det_logdet_slogdet_xpu_float64`. I reported the observation to MKL and it turned out to be a but, but it should have been fixed by now. Nonetheless, I would keep `at::zeros` as it doesn't introduce any noticeable overhead.

Agree with you.

pbielak commented:
CI now uses DLE2026.1 / MKL 2026.1 - rebasing and merging

pbielak commented:
@torchxpubot merge -f "CI failures are unrelated to PR changes"
